exceptions in SftpFileSystemProvider on Files.delete( Path p)

### Version

sshd-sftp 2.9.2

### Bug description

While using a Path on SftpFileSystemProvider:

Exceptions thrown in on Files.delete( Path p), when deleting symbolic links.

  * When deleting a symbolic links to non-existing file.
  * When deleting a symbolic link to existing directory
  * When deleting a symbolic link to existing file
 

### Actual behavior

probably the links are followed, and the link is handled like the linkdestination. this causes exceptions.

### Expected behavior

No exception but deletion of  the symbolic link.


### Relevant log output

_No response_

### Other information

Server OpenSSH_8.9p1 Ubuntu-3ubuntu0.1

```
Caused by: org.apache.sshd.sftp.common.SftpException: No such file
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.throwStatusException(AbstractSftpClient.java:217)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkResponseStatus(AbstractSftpClient.java:212)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkResponseStatus(AbstractSftpClient.java:186)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkCommandStatus(AbstractSftpClient.java:164)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.rmdir(AbstractSftpClient.java:817)
	at org.apache.sshd.sftp.client.fs.SftpFileSystem$Wrapper.rmdir(SftpFileSystem.java:372)
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.delete(SftpFileSystemProvider.java:591)
	at java.base/java.nio.file.Files.delete(Files.java:1152)
	at nl.qcg.br.releaser.util.FileUtil.fixedFileDelete(FileUtil.java:135)
	at nl.qcg.br.releaser.cmds.comp.PathCopier.lambda$remove$3(PathCopier.java:181)
	... 19 common frames omitted

```

````
Caused by: java.nio.file.NoSuchFileException: usr/share/icons/ubuntu-mono-light/stock/22/avatar-default.svg
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.resolveRemoteFileAttributes(SftpFileSystemProvider.java:994)
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.lambda$readRemoteAttributes$0(SftpFileSystemProvider.java:969)
	at org.apache.sshd.sftp.client.impl.SftpPathImpl.withAttributeCache(SftpPathImpl.java:120)
	at org.apache.sshd.sftp.client.impl.SftpPathImpl.withAttributeCache(SftpPathImpl.java:140)
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.readRemoteAttributes(SftpFileSystemProvider.java:969)
	at org.apache.sshd.sftp.client.impl.AbstractSftpFileAttributeView.readRemoteAttributes(AbstractSftpFileAttributeView.java:71)
	at org.apache.sshd.sftp.client.fs.SftpPosixFileAttributeView.readAttributes(SftpPosixFileAttributeView.java:52)
	at org.apache.sshd.sftp.client.fs.SftpPosixFileAttributeView.readAttributes(SftpPosixFileAttributeView.java:40)
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.checkAccess(SftpFileSystemProvider.java:826)
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.delete(SftpFileSystemProvider.java:581)
	at java.base/java.nio.file.Files.delete(Files.java:1152)
	at nl.qcg.br.releaser.cmds.comp.PathCopier.lambda$remove$3(PathCopier.java:180)
	... 19 common frames omitted
Caused by: org.apache.sshd.sftp.common.SftpException: No such file
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.throwStatusException(AbstractSftpClient.java:217)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkAttributesResponse(AbstractSftpClient.java:305)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkAttributes(AbstractSftpClient.java:283)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.stat(AbstractSftpClient.java:997)
	at org.apache.sshd.sftp.client.fs.SftpFileSystem$Wrapper.stat(SftpFileSystem.java:412)
	at org.apache.sshd.sftp.client.fs.SftpFileSystemProvider.resolveRemoteFileAttributes(SftpFileSystemProvider.java:981)
	... 30 common frames omitted
````

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -35,6 +35,7 @@
 * [GH-298](https://github.com/apache/mina-sshd/issues/298) Server side heartbeat not working.
 * [GH-300](https://github.com/apache/mina-sshd/issues/300) Read the channel id in `SSH_MSG_CHANNEL_OPEN_CONFIRMATION` as unsigned int.
 * [GH-313](https://github.com/apache/mina-sshd/issues/313) Log exceptions in the SFTP subsystem before sending a failure status reply.
+* [GH-325](https://github.com/apache/mina-sshd/issues/325) SftpFileSystemProvider: fix deletions of symlinks through Files.delete().
 
 
 * [SSHD-1295](https://issues.apache.org/jira/browse/SSHD-1295) Fix cancellation of futures and add options to cancel futures on time-outs.

--- a/sshd-sftp/src/main/java/org/apache/sshd/sftp/client/fs/SftpFileSystemProvider.java
+++ b/sshd-sftp/src/main/java/org/apache/sshd/sftp/client/fs/SftpFileSystemProvider.java
@@ -601,15 +601,16 @@ public class SftpFileSystemProvider extends FileSystemProvider {
     @Override
     public void delete(Path path) throws IOException {
         SftpPath p = toSftpPath(path);
-        checkAccess(p, AccessMode.WRITE);
 
         SftpFileSystem fs = p.getFileSystem();
         if (log.isDebugEnabled()) {
             log.debug("delete({}) {}", fs, path);
         }
-
+        if (fs.isReadOnly()) {
+            throw new AccessDeniedException("Filesystem is read-only: " + path.toString());
+        }
+        BasicFileAttributes attributes = readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
         try (SftpClient sftp = fs.getClient()) {
-            BasicFileAttributes attributes = readAttributes(path, BasicFileAttributes.class);
             if (attributes.isDirectory()) {
                 sftp.rmdir(path.toString());
             } else {

--- a/sshd-sftp/src/test/java/org/apache/sshd/sftp/client/fs/SftpFileSystemTest.java
+++ b/sshd-sftp/src/test/java/org/apache/sshd/sftp/client/fs/SftpFileSystemTest.java
@@ -35,6 +35,8 @@ import java.nio.file.FileStore;
 import java.nio.file.FileSystem;
 import java.nio.file.FileSystems;
 import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
 import java.nio.file.StandardOpenOption;
 import java.nio.file.attribute.BasicFileAttributes;
@@ -84,6 +86,7 @@ import org.apache.sshd.util.test.CoreTestSupportUtils;
 import org.hamcrest.BaseMatcher;
 import org.hamcrest.Description;
 import org.hamcrest.MatcherAssert;
+import org.junit.Assume;
 import org.junit.Before;
 import org.junit.FixMethodOrder;
 import org.junit.Test;
@@ -111,11 +114,7 @@ public class SftpFileSystemTest extends AbstractSftpFilesSystemSupport {
 
     @Test
     public void testFileSystem() throws Exception {
-        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(),
-                MapBuilder.<String, Object> builder()
-                        .put(SftpModuleProperties.READ_BUFFER_SIZE.getName(), IoUtils.DEFAULT_COPY_SIZE)
-                        .put(SftpModuleProperties.WRITE_BUFFER_SIZE.getName(), IoUtils.DEFAULT_COPY_SIZE)
-                        .build())) {
+        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(), defaultOptions())) {
             assertTrue("Not an SftpFileSystem", fs instanceof SftpFileSystem);
             testFileSystem(fs, ((SftpFileSystem) fs).getVersion());
         }
@@ -128,11 +127,7 @@ public class SftpFileSystemTest extends AbstractSftpFilesSystemSupport {
                 getCurrentTestName());
         CommonTestSupportUtils.deleteRecursive(lclSftp);
 
-        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(),
-                MapBuilder.<String, Object> builder()
-                        .put(SftpModuleProperties.READ_BUFFER_SIZE.getName(), IoUtils.DEFAULT_COPY_SIZE)
-                        .put(SftpModuleProperties.WRITE_BUFFER_SIZE.getName(), IoUtils.DEFAULT_COPY_SIZE)
-                        .build())) {
+        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(), defaultOptions())) {
             assertTrue("Not an SftpFileSystem", fs instanceof SftpFileSystem);
             Path parentPath = targetPath.getParent();
             Path clientFolder = lclSftp.resolve("client");
@@ -158,6 +153,140 @@ public class SftpFileSystemTest extends AbstractSftpFilesSystemSupport {
                 assertEquals("Mismatched data at " + i, (byte) (i - 5), data[i]);
                 assertEquals("Mismatched data at " + (i + buf.length), (byte) (i - 5), data[i + buf.length]);
             }
+        }
+    }
+
+    @Test // See GH-325
+    public void testDeleteLink() throws Exception {
+        // This test creates symbolic links.
+        Assume.assumeFalse(OsUtils.isWin32());
+        Path targetPath = detectTargetFolder();
+        Path lclSftp = CommonTestSupportUtils.resolve(targetPath, SftpConstants.SFTP_SUBSYSTEM_NAME, getClass().getSimpleName(),
+                getCurrentTestName());
+        CommonTestSupportUtils.deleteRecursive(lclSftp);
+
+        List<Path> toRemove = new ArrayList<>();
+        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(), defaultOptions())) {
+            assertTrue("Not an SftpFileSystem", fs instanceof SftpFileSystem);
+            Path parentPath = targetPath.getParent();
+            Path clientFolder = lclSftp.resolve("client");
+            assertHierarchyTargetFolderExists(clientFolder);
+            Path localFile = clientFolder.resolve("file.txt");
+            Files.write(localFile, "Hello".getBytes(StandardCharsets.UTF_8));
+            toRemove.add(localFile);
+            Path existingSymlink = clientFolder.resolve("existing.txt");
+            Files.createSymbolicLink(existingSymlink, localFile);
+            toRemove.add(existingSymlink);
+
+            String remExistingLink = CommonTestSupportUtils.resolveRelativeRemotePath(parentPath, existingSymlink);
+            Path remoteExistingLink = fs.getPath(remExistingLink);
+            assertHierarchyTargetFolderExists(remoteExistingLink.getParent());
+            assertTrue(Files.exists(remoteExistingLink));
+            assertTrue(Files.exists(remoteExistingLink, LinkOption.NOFOLLOW_LINKS));
+            assertTrue(Files.isSymbolicLink(remoteExistingLink));
+            Files.delete(remoteExistingLink);
+            assertTrue(Files.exists(localFile));
+            assertFalse(Files.exists(existingSymlink, LinkOption.NOFOLLOW_LINKS));
+            assertFalse(Files.exists(remoteExistingLink, LinkOption.NOFOLLOW_LINKS));
+        } finally {
+            for (Path p : toRemove) {
+                Files.deleteIfExists(p);
+            }
+        }
+    }
+
+    @Test // See GH-325
+    public void testDeleteNonexistingLink() throws Exception {
+        // This test creates symbolic links.
+        Assume.assumeFalse(OsUtils.isWin32());
+        Path targetPath = detectTargetFolder();
+        Path lclSftp = CommonTestSupportUtils.resolve(targetPath, SftpConstants.SFTP_SUBSYSTEM_NAME, getClass().getSimpleName(),
+                getCurrentTestName());
+        CommonTestSupportUtils.deleteRecursive(lclSftp);
+
+        List<Path> toRemove = new ArrayList<>();
+        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(), defaultOptions())) {
+            assertTrue("Not an SftpFileSystem", fs instanceof SftpFileSystem);
+            Path parentPath = targetPath.getParent();
+            Path clientFolder = lclSftp.resolve("client");
+            assertHierarchyTargetFolderExists(clientFolder);
+            Path nonExistingSymlink = clientFolder.resolve("nonexisting.txt");
+            Files.createSymbolicLink(nonExistingSymlink, clientFolder.resolve("gone.txt"));
+            toRemove.add(nonExistingSymlink);
+
+            String remNonExistingLink = CommonTestSupportUtils.resolveRelativeRemotePath(parentPath, nonExistingSymlink);
+            Path remoteNonExistingLink = fs.getPath(remNonExistingLink);
+            assertFalse(Files.exists(remoteNonExistingLink));
+            assertTrue(Files.exists(remoteNonExistingLink, LinkOption.NOFOLLOW_LINKS));
+            assertTrue(Files.isSymbolicLink(remoteNonExistingLink));
+            Files.delete(remoteNonExistingLink);
+            assertFalse(Files.exists(nonExistingSymlink, LinkOption.NOFOLLOW_LINKS));
+            assertFalse(Files.exists(remoteNonExistingLink, LinkOption.NOFOLLOW_LINKS));
+        } finally {
+            for (Path p : toRemove) {
+                Files.deleteIfExists(p);
+            }
+        }
+    }
+
+    @Test // See GH-325
+    public void testDeleteDirectoryLink() throws Exception {
+        // This test creates symbolic links.
+        Assume.assumeFalse(OsUtils.isWin32());
+        Path targetPath = detectTargetFolder();
+        Path lclSftp = CommonTestSupportUtils.resolve(targetPath, SftpConstants.SFTP_SUBSYSTEM_NAME, getClass().getSimpleName(),
+                getCurrentTestName());
+        CommonTestSupportUtils.deleteRecursive(lclSftp);
+
+        List<Path> toRemove = new ArrayList<>();
+        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(), defaultOptions())) {
+            assertTrue("Not an SftpFileSystem", fs instanceof SftpFileSystem);
+            Path parentPath = targetPath.getParent();
+            Path clientFolder = lclSftp.resolve("client");
+            assertHierarchyTargetFolderExists(clientFolder);
+            Path directory = clientFolder.resolve("subdir");
+            Files.createDirectory(directory);
+            toRemove.add(directory);
+            Path directorySymlink = clientFolder.resolve("dirlink");
+            Files.createSymbolicLink(directorySymlink, directory);
+            toRemove.add(directorySymlink);
+
+            String remDirectoryLink = CommonTestSupportUtils.resolveRelativeRemotePath(parentPath, directorySymlink);
+            Path remoteDirectoryLink = fs.getPath(remDirectoryLink);
+            assertTrue(Files.isDirectory(remoteDirectoryLink));
+            assertTrue(Files.exists(remoteDirectoryLink, LinkOption.NOFOLLOW_LINKS));
+            assertFalse(Files.isDirectory(remoteDirectoryLink, LinkOption.NOFOLLOW_LINKS));
+            assertTrue(Files.isSymbolicLink(remoteDirectoryLink));
+            Files.delete(remoteDirectoryLink);
+            assertTrue(Files.isDirectory(directory));
+            assertFalse(Files.exists(directorySymlink, LinkOption.NOFOLLOW_LINKS));
+            assertFalse(Files.exists(remoteDirectoryLink, LinkOption.NOFOLLOW_LINKS));
+        } finally {
+            for (Path p : toRemove) {
+                Files.deleteIfExists(p);
+            }
+        }
+    }
+
+    @Test // See GH-325
+    public void testDeleteNonexistingFile() throws Exception {
+        Path targetPath = detectTargetFolder();
+        Path lclSftp = CommonTestSupportUtils.resolve(targetPath, SftpConstants.SFTP_SUBSYSTEM_NAME, getClass().getSimpleName(),
+                getCurrentTestName());
+        CommonTestSupportUtils.deleteRecursive(lclSftp);
+
+        try (FileSystem fs = FileSystems.newFileSystem(createDefaultFileSystemURI(), defaultOptions())) {
+            assertTrue("Not an SftpFileSystem", fs instanceof SftpFileSystem);
+            Path parentPath = targetPath.getParent();
+            Path clientFolder = lclSftp.resolve("client");
+            assertHierarchyTargetFolderExists(clientFolder);
+
+            String doesNotExist = CommonTestSupportUtils.resolveRelativeRemotePath(parentPath,
+                    clientFolder.resolve("neverExisted.txt"));
+            Path neverExisted = fs.getPath(doesNotExist);
+            assertFalse(Files.exists(neverExisted));
+            assertFalse(Files.deleteIfExists(neverExisted));
+            assertThrows(NoSuchFileException.class, () -> Files.delete(neverExisted));
         }
     }
 

